### PR TITLE
LibJS: Fix toFixed() implementation for real this time :^)

### PR DESCRIPTION
**Pull request by ben-abraham**
This resolves the test262 test that was broken in my last PR. Also brings execution basically perfectly in line with modern browsers.
There is still one failing `toFixed()` test, but it's actually on the toString() impl here being TOO accurate (*sigh*)

```js
assert.sameValue((1000000000000000128).toString(), "1000000000000000100");
assert.sameValue((1000000000000000128).toFixed(0), "1000000000000000128");
```

Acid3 test behavior still working:
image

Previously broken test-262 case:
image

Comparison of worst-case-scenario to FF - exact match:
image

### CODE REVIEW

**Comment by ben-abraham** [comment]
I changed the import from `<AK/math.h>` to just `<math.h>` as the linter suggested, but it fails the build now, only on the server. It still builds and runs fine locally.

**Comment by ben-abraham** [comment]
Extra note: Formatter is still broken for float/double over U64MAX, will just always return "0.000000" This code could be used to fix it, but that would require the `BigInteger` classes from `LibCrypto` to do so. Maybe it makes sense to have AK bigint primitives?

File: Userland/Libraries/LibM/math.h [hunk 1]
```diff
@@ -16,6 +16,57 @@
...
+    static const int exponent_bits = 8;
+    static const unsigned exponent_max = 255;
+    struct {
+        unsigned long long mantissa : 23;
+        unsigned exponent : 8;
+        unsigned sign : 1;
+    };
+    float d;
+};
+
```
Comment (BertalanD) [inline Userland/Libraries/LibM/math.h:69]: This is going to make it impossible to use `math.h` in ports written in C.

As `FloatExtractor` is an implementation detail not specified by the C standard, please move it to a separate header. It might make sense to put it in `AK/Math.h`.

**Comment by linusg** [comment]
> I changed the import from `<AK/math.h>` to just `<math.h>` as the linter suggested, but it fails the build now, only on the server. It still builds and runs fine locally.

The linter does no such thing. There's one that requires to not use LibM/math.h includes, but that's different. The reason for this failure is that math.h is provided by the host system, so the lagom build doesn't find these newly added templates.
The solution is to do what Daniel suggests.

File: Userland/Libraries/LibM/math.h (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibM/math.h:69]: Let's not fill `AK/Math.h` with random things, more files are a thing, let's use them, `AK/FloatExtractor.h` is a perfectly fine file name.

**Comment by ben-abraham** [comment]
> > I changed the import from `<AK/math.h>` to just `<math.h>` as the linter suggested, but it fails the build now, only on the server. It still builds and runs fine locally.
> 
> The linter does no such thing. There's one that requires to not use LibM/math.h includes, but that's different. The reason for this failure is that math.h is provided by the host system, so the lagom build doesn't find these newly added templates. The solution is to do what Daniel suggests.

My bad, this is what I was thinking about, not AK. Moving to own header file as suggested.

File: Userland/Libraries/LibM/math.cpp [hunk 2]
```diff
@@ -56,57 +57,6 @@ enum class RoundingMode {
     ToEven = FE_TONEAREST
 };
 
-template<typename T>
```
Comment (linusg) [inline Userland/Libraries/LibM/math.cpp:59]: Can you put this code move in a separate commit, explaining why it's necessary? (standalone use of FloatExtractor)

File: Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp [hunk 3]
```diff
@@ -318,6 +344,22 @@ void UnsignedBigInteger::set_bit_inplace(size_t bit_index)
     m_cached_hash = 0;
 }
 
+void UnsignedBigInteger::unset_bit_inplace(size_t bit_index)
+{
+    const size_t word_index = bit_index / UnsignedBigInteger::BITS_IN_WORD;
+    const size_t inner_word_index = bit_index % UnsignedBigInteger::BITS_IN_WORD;
```
Comment (linusg) [inline Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp:350]: East const please

File: Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp [hunk 4]
```diff
@@ -318,6 +344,22 @@ void UnsignedBigInteger::set_bit_inplace(size_t bit_index)
...
+
+    m_words.ensure_capacity(word_index + 1);
+
+    for (size_t i = length(); i <= word_index; ++i) {
+        m_words.unchecked_append(0);
+    }
```
Comment (linusg) [inline Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp:356]: No curly braces needed

File: Userland/Libraries/LibJS/Runtime/NumberPrototype.h [hunk 5]
```diff
@@ -6,6 +6,8 @@
 
 #pragma once
 
+#include <LibCrypto/BigInt/SignedBigInteger.h>
+#include <LibCrypto/NumberTheory/ModularFunctions.h>
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/NumberPrototype.h:10]: Please add these includes to the cpp file instead if the header doesn't use them

**Review by linusg — CHANGES_REQUESTED** [review]

**Review by linusg — COMMENTED** [review]
Please also wrap commit messages at 72 chars, not less.

File: Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp [hunk 6]
```diff
@@ -318,6 +344,22 @@ void UnsignedBigInteger::set_bit_inplace(size_t bit_index)
     m_cached_hash = 0;
 }
 
+void UnsignedBigInteger::unset_bit_inplace(size_t bit_index)
```
Comment (IdanHo) [inline Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp:347]: the nomenclature is usually set and clear, so `clear_bit_inplace` would be a bit nicer

File: Userland/Libraries/LibCrypto/BigInt/SignedBigInteger.cpp [hunk 7]
```diff
@@ -16,6 +19,38 @@ SignedBigInteger SignedBigInteger::import_data(const u8* ptr, size_t length)
...
+    int sign = extractor.sign;
+    auto unbiased_exponent = extractor.exponent - Extractor::exponent_bias;
+    unsigned long long significand = extractor.mantissa;
+
+    significand |= 1ull << Extractor::mantissa_bits;
+
+    UnsignedBigInteger number_value;
+    if (unbiased_exponent > Extractor::mantissa_bits) {
+        auto scalar = Crypto::NumberTheory::Power(UnsignedBigInteger(2), UnsignedBigInteger(unbiased_exponent - Extractor::mantissa_bits));
+        number_value = UnsignedBigInteger::create_from(significand).multiplied_by(multiplier).multiplied_by(scalar);
```
Comment (IdanHo) [inline Userland/Libraries/LibCrypto/BigInt/SignedBigInteger.cpp:38]: can we use UnsignedBigInteger::shift_left() instead?

File: Userland/Libraries/LibCrypto/BigInt/SignedBigInteger.h [hunk 8]
```diff
@@ -60,6 +61,11 @@ class SignedBigInteger {
         }
         return SignedBigInteger { UnsignedBigInteger::create_from(unsigned_value), sign };
     }
+    static SignedBigInteger create_from_double(double value)
+    {
+        return create_from_double_with_multiplier(value, 1);
```
Comment (IdanHo) [inline Userland/Libraries/LibCrypto/BigInt/SignedBigInteger.h:66]: multiplication by 1 is currently O(N) can we avoid this somehow? (and/or add a fast path to UnsignedBigInteger::multiplied_by & divided_by)

**Comment by ben-abraham** [comment]
Been doing some digging on-and-off for where exactly the breakdown is with the string representation. I think it ultimately comes down to the fact that internally [we use](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibJS/Runtime/Temporal/Duration.h#L40) doubles for all of the `Duration` prototypes, while the [documentation](https://tc39.es/proposal-temporal/#sec-properties-of-temporal-duration-instances) says to use integer numbers, I presume so nanoseconds can hold values greater than `U64MAX`. For this same reason, we can't simply `round()` all results either, as there wouldn't be enough bits of precision in some cases to round to a single integer value (this crossover point is at `(2^53) - 1` or `18,014,398,509,481,983ns` ~= 208 Days (with `U64MAX` ~= 584 Years.)

I would think switching to all `BigInteger` would be appropriate....but also overkill, so maybe do just `u64` for all of the fields and then BigInteger alone for the nanoseconds field?

If that were done, I think that would fix most/all of the rounding errors that are causing tests to fail. (Such as 0.5 being printed when a 0 is expected on an integer field)
